xsd/xsd.go

```diff
@@ -38,6 +39,107 @@ type Type interface {
 	isType()
 }
 
+func TypesEqual(a, b Type) bool {
+	switch a := a.(type) {
+	case *SimpleType:
+		switch b := b.(type) {
+		case *SimpleType:
+			return simpleTypeEqual(a, b)
+		default:
+			return false
+		}
+	case *ComplexType:
+		switch b := b.(type) {
+		case *ComplexType:
+			return complexTypeEqual(a, b)
+		default:
+			return false
+		}
+	case Builtin:
+		switch b := b.(type) {
+		case Builtin:
+			return a == b
+		default:
+			return false
+		}
+	case linkedType:
+		// FIXME: implement
+		return false
+	default:
+		return false
+	}
+}
+
+func unionsEqual(a []Type, b []Type) bool {
+	// TODO: handle out of order
+	if len(a) != len(b) {
+		// TODO: handle duplicate union entries making one list longer
+		return false
+	}
+	for i := range a {
+		if !TypesEqual(a[i], b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func restrictionsEqual(a Restriction, b Restriction) bool {
+	if len(a.Enum) != len(b.Enum) {
+		// TODO: handle duplicate union entries making one list longer
+		return false
+	}
+	aEnum := make([]string, len(a.Enum))
+	bEnum := make([]string, len(b.Enum))
+	copy(aEnum, a.Enum)
+	copy(bEnum, b.Enum)
+	sort.Strings(aEnum)
+	sort.Strings(bEnum)
+	for i := range aEnum {
+		if aEnum[i] != bEnum[i] {
+			return false
+		}
+	}
+	if a.Min != b.Min || a.Max != b.Max ||
+		a.Length != b.Length || a.MinLength != b.MinLength || a.MaxLength != b.MaxLength ||
+		a.MinDate != b.MinDate || a.MaxDate != b.MaxDate ||
+		a.Precision != b.Precision ||
+		a.TotalDigits != b.TotalDigits {
+		return false
+	}
+
+	if (a.Pattern != nil && b.Pattern == nil) ||
+		(a.Pattern == nil && b.Pattern != nil) ||
+		(a.Pattern != nil && a.Pattern.String() != b.Pattern.String()) {
```

Comment:
```suggestion
		(a.Pattern != nil && b.Pattern != nil &&  a.Pattern.String()  != b.Pattern.String()) {
```

Reply:
What if only one of them has a pattern?

Reply:
but calling `.Pattern.String()` on a nil `Pattern` would panic right?

Reply:
correct

Reply:
yeah then you have to check if both a and b patterns a not nil

Reply:
Oh, sorry, I misread your suggestion. Only one pattern being nil would be caught by the first two checks.